srtcore/common.h

```diff
@@ -167,6 +167,270 @@ const int DEFAULT_LIVE_LATENCY = 120; // (mSec)
+
+    In order that this make sense there would be required an array
+    indexed by event id (just like a slot array m_Slots in CUDT),
+    where the "type distiller" function would be extracted and then
+    combined with the user-connected slot function this would call
+    it already with correct type. Note that also the ConnectSignal
+    function would have to get the signal id by template parameter,
+    not function parameter. For example:
+
+    m_parent->ConnectSignal<TEV_ACK>(SSLOT(updateOnSent));
+
+    in which updateOnSent would have to receive an appropriate type.
+    This has a disadvantage that you can't connect multiple signals
+    with different argument types to the same slot, you'd have to
+    make slot wrappers to translate arguments.
+
+    It seems that a better idea would be to create binders that would
+    translate the argument from EventArg to the correct type according
+    to the rules imposed by particular event id. But I'd not make it
+    until there's a green light on C++11 for SRT, so maybe in a far future.
+
+template <ETransmissionEvent type>
+class EventArgType;
+#define MAP_EVENT_TYPE(tev, tp) template<> class EventArgType<tev> { typedef tp type; }
+*/
+
+
+// The 'type' field wouldn't be even necessary if we
+// could use any method of extracting 'type' from 'type U::*' expression.
+template<> struct EventVariant::VariantFor<EventVariant::PACKET>
+{
+    typedef CPacket* type;
+    static type U::*field() {return &U::packet;}
+};
+
+template<> struct EventVariant::VariantFor<EventVariant::ACK>
+{
+    typedef uint32_t type;
+    static type U::*field() { return &U::ack; }
+};
+
+template<> struct EventVariant::VariantFor<EventVariant::STAGE>
+{
+    typedef ECheckTimerStage type;
+    static type U::*field() { return &U::stage; }
+};
+
+template<> struct EventVariant::VariantFor<EventVariant::INIT>
+{
+    typedef EInitEvent type;
+    static type U::*field() { return &U::init; }
+};
+
+// Sigh. The code must be C++03, C++11 and C++17 compliant.
+// There's std::mem_fun in C++03, but it's deprecated in C++11 and removed in
+// C++17.  There's std::function and std::bind, but only since C++11. No way to
+// define it any compatible way.  There were already problems with ref_t and
+// unique_ptr/auto_ptr, for which custom classes were needed.
+
+// This can be changed in future to use C++11 way, but only after C++03
+// compatibility is finally abaondoned. Until then, this stays with a custom class.
+
+class EventSlotBase
+{
+public:
+    virtual void emit(ETransmissionEvent tev, EventVariant var) = 0;
+    typedef void dispatcher_t(void* opaque, ETransmissionEvent tev, EventVariant var);
+
+    virtual ~EventSlotBase() {}
+};
+
+class SimpleEventSlot: public EventSlotBase
+{
+public:
+    void* opaque;
+    dispatcher_t* dispatcher;
+
+    SimpleEventSlot(void* op, dispatcher_t* disp): opaque(op), dispatcher(disp) {}
+
+    void emit(ETransmissionEvent tev, EventVariant var) ATR_OVERRIDE
```

Comment:
It's rather more something like `std::function` in C++11. This whole complicated mechanism is in order to make a creator of a new Smoother bind with the existing code easily, also when the same action must be taken for multiple events and vice versa.

And why these are passed by variants rather than requiring exact type from the handler, it's explained at the previous comment. This would have to be rewritten so that `ConnectSignal` gets the signal by template parameter and it's then associated with type signature. I didn't have the right idea to do it at that time, so I made it easily with variants. Variants would still be required, but value extraction would happen before calling the handler. This is something that can be changed in future.